manlinks.d/C/struct.7
.TH struct 7 2026-10-01
.SH NAME
.B struct
\- digest of concepts and practices
.SH DESCRIPTION
.PP
Structs in C are the foundation of C++ classes.  Structs can express
much of the same utility as a class.
.SH SYNOPSIS
.PP
.EX
/* Typedefs and Definitions */
.B "typedef struct address" \c
.IR address_t ;
.B "typedef struct person" \c
.IR person_t ;

.BI struct " address " \c
{
.RS 4
.BR "const char " *\c
.IR street ;
.\"
.BR "const char " *\c
.IR postcode ;
.RE
};

.BI struct " person " \c
{
.RS 4
.BR "const char " *\c
.IR name ;
.\"
.BR "struct address "\c
.IR address ;
.RE
};

/* Initialization */
.BI "struct address" " home " \c
.RI "= { \(dq" "12 Main St" "\(dq, \(dq" 10000 "\(dq };"

.BI person_t " Mama "\c
.RI "= { \(dq" "Mama Bear" "\(dq, { \(dq" "12 main St" "\(dq, \(dq" 10000 "\(dq }};"

.BI person_t " Papa "\c
= {
.RS 4
.RI . name " = \(dq" "Papa Bear" "\(dq,"
.RI . address " = { \(dq" "12 Main Street" "\(dq, \(dq" 10000 "\(dq }"
.RE
};

.BI person_t " Sister "\c
= {
.RS 4
.RI . name " = \(dq" "Sister Bear" "\(dq,"
.RI . address . postcode " = \(dq" 10000 \(dq
.RE
};


.BI person_t " Brother "\c
= {
.RS 4
.RI . name " = \(dq" "Brother Bear" "\(dq,"
.RI . address " = " Mama . address
.RE
};

/* Compound Literal */
.BI address_t " neighbor " \c
.RB "= (" address_t "){ \(dq" \c
.IR "100 Main St" "\(dq, \(dq" 10000 "\(dq };"
.\"
.B address_t \c
.RI * favorite \c
.RB " = &(" address_t ) \c
.RI "{ \(dq" "100 Main St" "\(dq, \(dq" 10000 "\(dq };"
.\" 
.BI person_t " Baby " \c
.RB "= (" person_t ){
.RS 4
.RI . name " = \(dq" "Baby Bear" \(dq,
.RI . address " = " \c
.RB ( address_t ){
.RS 4
.RI . street " = \(dq" "100 Main St" "\(dq,"
.RI . postcode " = \(dq" 10000 "\(dq"
.RE
}
.RE
};

/* Assignments */
.B person_t \c
.IR "friend" ;
.\"
/* from another instance: */
.IR friend " = " Sister ";"

/* from compound literal
.IR friend " = ("\c
.BR person_t ") {"
.RS 4
.RI "\(dq" "Tammy Bear" "\(dq,"
{ \c
.RI "\(dq" "24 Elm St" "\(dq," \c
.RI "\(dq" "10000" "\(dq " \c
}
.RE
};

.SH DESCRIPTION
.PP
Structs are an outlier in the C programming language.  While most
things in C are simple and direct, a struct is a composite type with
unusual (for C) rules initialization and assignment.
.PP
Typically, struct members are accessed individually for assignment
or for their value.  This page is about less well-known methods for
preparing a struct by inialization and assignment.
.SS INITIALIZATION
.PP
Braces, AKA curly-brackets, are used to initialize a struct instance
in a single statement.  During initialization, the member can be
assigned by position or by name.
.PP
.EX
/* by an Initialization List (anonymous, by position) */
.BI "struct address" " home " \c
= {
.RS 4
.RI \(dq "12 Main St" \(dq,
.RI \(dq "55555" \(dq
.RE
};

/* by a list of Designated Initializers (by name, random order) */
.BI "struct address" " home " \c
= {
.RS 4
.RI . postcode " = \(dq" "55555" \(dq,
.RI . street "   = \(dq" "12 Main St" \(dq
.RE
};
.EE
.PP
Uninitialized struct instances should be assumed to contain random
member values.  However, with an initialization statement, members
not specifically addressed by field initializers will be set to 0:
.PP
.EX
.BI "struct address" " home "\c
.RI "= { \(dq" "12 Main St" "\(dq };"
.\"
.B printf \c
.RI "(\(dq" "postcode = %s\(rsn" "\(dq, " home . postcode ");"

postcode = (null)
.EE
.SS COMPOUND LITERAL
.PP
A
.B compound literal
is an anonymous data item in memory, initialized with a type name
and an initialization list.  This can be used to create simple or
complex data types, including arrays and structs.
.PP
A struct instance can be set using a compound literal:
.PP
.EX
.B struct address\c
.IR " home" ;
.IR home " = ("\c
.BR "struct address" "){ "
.RS 4
.RI . postcode " = \(dq" "55555" \(dq,
.RI . street "   = \(dq" "12 Main St" \(dq
.RE
};
.PP
Since the result of a compound literal is an anonymous instance in
memory, you can also use the address of the instance.  For example,
pass the address of the compound literal as an argument to a function.


manlinks.d/C/struct.c
#include <stdio.h>

       /* Typedefs and Definitions */
       typedef struct address address_t;
       typedef struct person person_t;

       struct address {
           const char *street;
           const char *postcode;
       };

       struct person {
           const char *name;
           struct address address;
       };

void initializations(void)
{
       /* Initialization */
       struct address home = { "12 Main St", "10000" };

       person_t Mama = { "Mama Bear", { "12 main St", "10000" }};

       person_t Papa = {
           .name = "Papa Bear",
           .address = { "12 Main Street", "10000" }
       };

       person_t Sister = {
           .name = "Sister Bear",
           .address.postcode = "10000"
       };

       person_t Brother = {
           .name = "Brother Bear",
           .address = Mama.address
       };

       /* Compound Literal (cast to struct) */
       address_t neighbor = (address_t){ "100 Main St", "10000" };
       address_t *favorite = &(address_t){ "100 Main St", "10000" };
       person_t Baby = (person_t){
           .name = "Baby Bear",
           .address = (address_t){
               .street = "100 Main St",
               .postcode = "10000"
           }
       };
}

void assignments(void)
{
       /* Assignments */
       person_t friend;
       friend.name = "Tammy";
       friend.address = (struct address){ "24 Elm St", "10000" };
}


#ifdef STRUCT_MAIN
typedef struct person_plus person_plus_t;
struct person_plus {
   struct person base;
   int           age;
   const char    *profession;
};

void show_person(const person_t *person)
{
   printf("name:        %s\n"
          "address:\n"
          "   street:   %s\n"
          "   postcode: %s\n",
          person->name, person->address.street, person->address.postcode);
}

void show_person_plus(const person_plus_t *person)
{
   show_person((person_t*)person);
   printf("age:         %d\n"
          "profession:  %s\n",
          person->age, person->profession);
}

void demo_person_plus(void)
{
   person_plus_t per = (person_plus_t){"Joe", {"45 Skid Row", "02020"}, 29, "rabble-rouser"};
   person_t *dude = (person_t*)&per;

   show_person_plus((person_plus_t*)dude);
}

void demo_named_fields(void)
{
   person_plus_t per = (person_plus_t){
      .base.name = "Roger",
      .age = 35,
      .profession = "director"
   };

   show_person_plus(&per);
}



int main(int argc, const char **argv)
{
   initializations();
   assignments();

   demo_person_plus();
   demo_named_fields();

   return 0;
}



#endif

/* Local Variables:        */
/* compile-command: " gcc \*/
/* -DSTRUCT_MAIN          \*/
/* -ggdb                  \*/
/* -o struct struct.c"    \*/
/* End:                    */
